reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,505 @@
+            if (retrievableEstimates.containsKey(resolvable)) return retrievableEstimates.get(resolvable);
+            else if (inferrableEstimates == null) {
+                assert this.initializationStatus == InitializationStatus.IN_PROGRESS || this.initializationStatus == InitializationStatus.RESET;
+                return new ArrayList<>();
+            } else {
+                return inferrableEstimates.get(resolvable);
+            }
+        }
+        // </editor-fold>
+
+        // <editor-fold desc="query">
+        public long estimateAllAnswers() {
+            assert this.fullAnswerCount >= 0;
+            return this.fullAnswerCount;
+        }
+
+        public long estimateAnswers(Set<Variable> variableFilter, Set<Resolvable<?>> includedResolvables) {
+            if (this.initializationStatus == InitializationStatus.IN_PROGRESS) {
+                answerCountEstimator.reportInitializationCycle(this.conjunction);
+            }
+
+            List<LocalEstimate> includedEstimates = iterate(includedResolvables)
+                    .flatMap(resolvable -> iterate(multivarEstimate(resolvable)))
+                    .toList();
+            Map<Variable, LocalEstimate> costCover = computeCheapestEstimateCoverForVariables(variableFilter, includedEstimates);
+            long ret = costOfEstimateCover(variableFilter, costCover);
+            assert ret > 0;             // Flag in tests if it happens.
+            return Math.max(ret, 1);    // Don't do stupid stuff in prod when it happens.
+        }
+
+        private Map<Variable, LocalEstimate> computeCheapestEstimateCoverForVariables(Set<Variable> variableFilter, List<LocalEstimate> includedEstimates) {
+            // Does a greedy set cover
+            Map<Variable, LocalEstimate> costCover = new HashMap<>(unaryEstimateCover);
+            includedEstimates.sort(Comparator.comparing(x -> x.answerEstimate(variableFilter)));
+            for (LocalEstimate estimate : includedEstimates) {
+                Set<Variable> filteredVariablesInEstimate = estimate.variables.stream()
+                        .filter(variableFilter::contains).collect(Collectors.toSet());
+
+                long currentCostToCover = costOfEstimateCover(filteredVariablesInEstimate, costCover);
+                if (currentCostToCover > estimate.answerEstimate(filteredVariablesInEstimate)) {
+                    filteredVariablesInEstimate.forEach(v -> costCover.put(v, estimate));
+                }
+            }
+            return costCover;
+        }
+
+        private static long costOfEstimateCover(Set<Variable> variablesToConsider, Map<Variable, LocalEstimate> coverMap) {
+            Set<LocalEstimate> subsetCoveredBy = coverMap.keySet().stream().filter(variablesToConsider::contains)
+                    .map(coverMap::get).collect(Collectors.toSet());
+            return subsetCoveredBy.stream().map(estimate -> estimate.answerEstimate(variablesToConsider)).reduce(1L, (x, y) -> x * y);
+        }
```

Comment:
Computes the cost of covering a set of variables given a (partial) estimate cover.